src/Compilers/Core/Portable/Compilation/Compilation.cs

```diff
@@ -2449,99 +2449,36 @@ internal void EnsureAnonymousTypeTemplates(CancellationToken cancellationToken)
                 }
 
-                Func<Stream> getPortablePdbStream;
-                if (moduleBeingBuilt.DebugInformationFormat == DebugInformationFormat.PortablePdb && pdbStreamProvider != null)
-                {
-                    getPortablePdbStream = () =>
-                    {
-                        if (metadataDiagnostics.HasAnyErrors())
-                        {
-                            return null;
-                        }
-
-                        portablePdbStream = pdbStreamProvider.GetOrCreateStream(metadataDiagnostics);
-                        Debug.Assert(portablePdbStream != null || metadataDiagnostics.HasAnyErrors());
-                        return portablePdbStream;
-                    };
-                }
-                else
-                {
-                    getPortablePdbStream = null;
-                }
 
-                // PROTOTYPE(strongname): simplify this code
                 Func<Stream> getPeStream = () =>
                 {
-                    if (metadataDiagnostics.HasAnyErrors())
-                    {
-                        return null;
-                    }
-
-                    peStream = peStreamProvider.GetOrCreateStream(metadataDiagnostics);
-                    if (peStream == null)
-                    {
-                        Debug.Assert(metadataDiagnostics.HasAnyErrors());
-                        return null;
-                    }
-
-                    // If the current strong name provider is the Desktop version, signing can only be done to on-disk files. 
-                    // If this binary is configured to be signed, create a temp file, output to that
-                    // then stream that to the stream that this method was called with. Otherwise output to the
-                    // stream that this method was called with.
-                    Stream retStream;
-                    if (!metadataOnly && IsRealSigned && this.Options.StrongNameProvider.Capability == SigningCapability.SignsStream)
-                    {
-                        Debug.Assert(Options.StrongNameProvider != null);
-
-                        try
-                        {
-                            signingInputStream = Options.StrongNameProvider.CreateInputStream();
-                        }
-                        catch (IOException e)
-                        {
-                            throw new Cci.PeWritingException(e);
-                        }
-
-                        retStream = signingInputStream;
-                    }
-                    else
-                    {
-                        signingInputStream = null;
-                        retStream = peStream;
-                    }
-
-                    return retStream;
+                    Stream ret;
+                    (peStream, signingInputStream, ret) = GetPeStream(metadataDiagnostics, peStreamProvider, metadataOnly);
```

Comment:
Is it necessary to return three values? It looks like we're only using the third value.

Reply:
The worse part of this method is the crazy things that happen inside these closures.

This closure (like the one that I replaced), mutates peStream and signingInputStream in the outer scope when it is run.